code/datums/components/trader/trader.dm

```diff
@@ -0,0 +1,413 @@
+	if(buyer_will_buy != "Yes" || !can_trade(customer))
+		return
+	trader.face_atom(customer)
+	if(!spend_buyer_offhand_money(customer, product_info[TRADER_PRODUCT_INFO_PRICE]))
+		trader.say(trader_data.return_trader_phrase(NO_CASH_PHRASE))
+		return
+	item_to_buy = new item_to_buy(get_turf(customer))
+	customer.put_in_hands(item_to_buy)
+	playsound(trader, trader_data.sell_sound, 50, TRUE)
+	log_econ("[item_to_buy] has been sold to [customer] (typepath used for product info; [item_to_buy.type]) by [trader] for [product_info[TRADER_PRODUCT_INFO_PRICE]] cash.")
+	product_info[TRADER_PRODUCT_INFO_QUANTITY] -= 1
+	trader.say(trader_data.return_trader_phrase(BUY_PHRASE))
+
+///Calculates the value of money in the hand of the buyer and spends it if it's sufficient
+/datum/component/trader/proc/spend_buyer_offhand_money(mob/customer, the_cost)
+	var/value = 0
+	var/obj/item/holochip/cash = customer.is_holding_item_of_type(/obj/item/holochip)
+	if(cash)
+		value += cash.credits
+	if((value >= the_cost) && cash)
+		return cash.spend(the_cost)
+	return FALSE //Purchase unsuccessful
+
+/**
+ * Tries to call sell_item on one of the customer's held items, if fail gives a chat message
+ *
+ * Gets both items in the customer's hands, and then tries to call sell_item on them, if both fail, he gives a chat message
+ * Arguments:
+ * * customer - (Mob REF) The mob trying to sell something
+ */
+/datum/component/trader/proc/try_sell(mob/customer)
+	if(!can_trade(customer))
+		return
+	var/mob/living/trader = parent
+	var/sold_item = FALSE
+	for(var/obj/item/an_item in customer.held_items)
+		if(sell_item(customer, an_item))
+			sold_item = TRUE
+			break
+	if(!sold_item && !can_trade(customer)) //don't talk if you are dead or fighting
+		trader.say(trader_data.return_trader_phrase(ITEM_REJECTED_PHRASE))
+
+
+/**
+ * Checks if an item is in the list of wanted items and if it is after a Yes/No radial returns generate_cash with the value of the item for the NPC
+ * Arguments:
+ * * customer - (Mob REF) The mob trying to sell something
+ * * selling - (Item REF) The item being sold
+ */
+/datum/component/trader/proc/sell_item(mob/customer, obj/item/selling)
+	var/mob/living/trader = parent
+	var/cost
+	if(!selling)
+		return FALSE
+	var/list/product_info
+	//Keep track of the typepath; rather mundane but it's required for correctly modifying the wanted_items
+	//should a product be sellable because even if it doesn't have a entry because it's a child of a parent that is present on the list
+	var/typepath_for_product_info
+	if(selling.type in wanted_items)
+		product_info = wanted_items[selling.type]
+		typepath_for_product_info = selling.type
+	else //Assume wanted_items is setup in the correct way; read wanted_items documentation for more info
+		for(var/typepath in wanted_items)
+			if(istype(selling, typepath))
+				product_info = wanted_items[typepath]
+				typepath_for_product_info = typepath
+				break
```

Comment:
this indentation is garish so let's add some stuff to make it better

```suggestion
		for(var/typepath in wanted_items)
			if(!istype(selling, typepath))
				continue

			product_info = wanted_items[typepath]
			typepath_for_product_info = typepath
			break
```

Reply:
That is indeed prettier